crates/agent/src/controllers/capture.rs

```diff
@@ -24,28 +24,35 @@ pub async fn update<C: ControlPlane>(
         return Ok(Some(NextRun::immediately()));
     }
 
-    let activate_next_run =
+    let activate_result =
         activation::update_activation(&mut status.activation, state, events, control_plane)
             .await
-            .with_retry(backoff_data_plane_activate(state.failures))?;
+            .with_retry(backoff_data_plane_activate(state.failures))
+            .map_err(Into::into);
 
-    // Return the publication error now, if there is one.
-    let _ = published?;
-
-    publication_status::update_notify_dependents(&mut status.publications, state, control_plane)
-        .await
-        .context("failed to notify dependents")?;
+    // Only notify dependents if our publication and activation were successful.
+    let notify_result = if published.is_ok() && activate_result.is_ok() {
+        publication_status::update_notify_dependents(&mut status.publications, state, control_plane)
+            .await
+            .context("failed to notify dependents")
+            .map(|_| None)
+    } else {
+        Ok(None)
+    };
 
     let ad_next = status
         .auto_discover
         .as_ref()
         .and_then(auto_discover::next_run);
     let periodic_next = periodic::next_periodic_publish(state);
-    Ok(NextRun::earliest([
-        ad_next,
-        periodic_next,
-        activate_next_run,
-    ]))
+
+    coalesce_results([
+        published.map(|_| None),
```

Comment:
This makes `Ok(bool)` into `Ok(None)`, right? that's all it does

Reply:
Yep, the publish result contains a boolean that says whether it did publish or not, and the next runs are handled by separate functions for each sub-status. Admittedly, this is not the clearest factoring, but it doesn't seem like a good idea to expand the scope of the refactor any more than necessary at this time.